crates/client/rpc/src/cache/mod.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use mc_storage::{OverrideHandle, StorageOverride};
+use mp_starknet::block::Block as StarknetBlock;
+use mp_starknet::storage::StarknetStorageSchemaVersion;
+use sc_service::SpawnTaskHandle;
+use sp_runtime::traits::Block as BlockT;
+use tokio::sync::{mpsc, oneshot};
+
+use crate::cache::lru_cache::LRUCache;
+
+mod lru_cache;
+
+enum StarknetDataCacheMessage<B: BlockT> {
+    RequestBlockByHash {
+        block_hash: B::Hash,
+        schema: StarknetStorageSchemaVersion,
+        response_tx: oneshot::Sender<Option<StarknetBlock>>,
+    },
+    FetchedBlockByHash {
+        block_hash: B::Hash,
+        block: Box<Option<StarknetBlock>>,
+    },
+}
+
+/// Manage LRU caches for block data and their transaction statuses.
+/// These are large and take a lot of time to fetch from the database.
+/// Storing them in an LRU cache will allow to reduce database accesses
+/// when many subsequent requests are related to the same blocks.
+pub struct StarknetDataCacheTask<B: BlockT>(mpsc::Sender<StarknetDataCacheMessage<B>>);
+
+/// Represents the metadata of the block we are looking for.
+struct BlockMetadata<B: BlockT> {
+    hash: B::Hash,
+    schema: StarknetStorageSchemaVersion,
+}
+
+/// Represent a pair of a cache and a data waitlist, used to communicate information between our
+/// internal methods.
+struct CacheWaitlist<B: BlockT, T> {
+    cache: LRUCache<<B as BlockT>::Hash, T>,
```

Comment:
not sure we need to use this hasher here tho, we should probably optimize for performance here w/ blake2